### PR TITLE
Threadpool suppresses subsequent runs if an exception is encountered.

### PR DESCRIPTION
**Pull request by AndrewGorton**
See http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate-java.lang.Runnable-long-long-java.util.concurrent.TimeUnit-

This wraps the heartbeating code with try/catch to ensure it can self-recover (otherwise we'd need to restart the app to restart heartbeating).

Note: The same problem exists with the processing threads - if they encounter an exception, subsequent runs will be suppressed. This is how some registers stop getting updated.

### CODE REVIEW

File: src/main/java/uk/gov/indexer/ExceptionFormatter.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+public class ExceptionFormatter {
+    public static String formatExceptionAsString(Exception e) {
+        StringWriter sw = new StringWriter();
+        PrintWriter pw = new PrintWriter(sw);
+        e.printStackTrace(pw);
+        return sw.toString();
+    }
```
Comment (AndrewGorton) [inline src/main/java/uk/gov/indexer/ExceptionFormatter.java:12]: Alternatives to this home-grown method include the [commons-lang.ExceptionUtils::getStackTrace() function](https://commons.apache.org/proper/commons-lang/javadocs/api-3.4/org/apache/commons/lang3/exception/ExceptionUtils.html#getStackTrace%28java.lang.Throwable%29).

**Comment by philandstuff** [comment]
Do we have examples of exceptions that were being thrown that this will handle appropriately?

**Comment by AndrewGorton** [comment]
Counterquestion: Are there any exceptions which should kill the heartbeat thread but leave the application running and processing records?

---

The [AWS API docs](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/cloudwatch/AmazonCloudWatchClient.html#putMetricData%28com.amazonaws.services.cloudwatch.model.PutMetricDataRequest%29) specify that Amazon may throw an [InternalServiceException](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/cloudwatch/model/InternalServiceException.html). But other things may also get thrown if people are reconfiguring things (eg IAM access denied, timeouts if the security groups are being reconfigured, etc).

**Comment by philandstuff** [comment]
After discussing we've decided to remove process heartbeating entirely from indexer, and use external process monitoring instead.

However, in the meantime, I think this is the Right Thing, so I'm merging.

**Pull request merged**